### PR TITLE
8265462: Handle multiple slots in the NSS Internal Module from SunPKCS11's Secmod

### PR DESCRIPTION
**Pull request by martinuy**
Hi,

Please find in this PR a proposal to fix JDK-8265462 [1].

With this fix, OpenJDK will only use the known slot IDs for the NSS Internal Module. If the NSS Internal Module has more slots (for example, as a result of an initialization sequence such as the one triggered from the libnsssysinit.so library), they will be ignored. The goal is to handle multiple-slots scenarios while keeping OpenJDK's previous behavior.

No regressions observed in the jdk/sun/security/pkcs11 tests category.

A new regression test was not added as part of this changeset because of its complexity. It would depend on a specific NSS configuration, or the NSS library would need to be mocked. I've done manual testing in my environment and JDK-8265462 [1] has further information about it.

Thanks,
Martin.-

--
[1] - https://bugs.openjdk.java.net/browse/JDK-8265462

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8265462](https://bugs.openjdk.java.net/browse/JDK-8265462): Handle multiple slots in the NSS Internal Module from SunPKCS11's Secmod

### Reviewers
 * [Valerie Peng](https://openjdk.java.net/census#valeriep) (@valeriepeng - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3661/head:pull/3661` \
`$ git checkout pull/3661`

Update a local copy of the PR: \
`$ git checkout pull/3661` \
`$ git pull https://git.openjdk.java.net/jdk pull/3661/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3661`

View PR using the GUI difftool: \
`$ git pr show -t 3661`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3661.diff">https://git.openjdk.java.net/jdk/pull/3661.diff</a>

</details>

### CODE REVIEW

File: src/jdk.crypto.cryptoki/share/native/libj2pkcs11/pkcs11t.h [hunk 1]
```diff
@@ -102,9 +102,14 @@ typedef CK_ULONG CK_NOTIFICATION;
 #define CKN_SURRENDER           0UL
 #define CKN_OTP_CHANGED         1UL
 
+typedef CK_ULONG CK_PROFILE_ID;
```
Comment (valeriepeng) [inline src/jdk.crypto.cryptoki/share/native/libj2pkcs11/pkcs11t.h:105]: There is already a typedef for CK_PROFILE_ID further down in this file.

File: src/jdk.crypto.cryptoki/share/native/libj2pkcs11/pkcs11t.h [hunk 2]
```diff
@@ -102,9 +102,14 @@ typedef CK_ULONG CK_NOTIFICATION;
...
+
 typedef CK_ULONG          CK_SLOT_ID;
 
 typedef CK_SLOT_ID CK_PTR CK_SLOT_ID_PTR;
+#define NETSCAPE_SLOT_ID 1
+#define PRIVATE_KEY_SLOT_ID 2
+#define FIPS_SLOT_ID 3
```
Comment (valeriepeng) [inline src/jdk.crypto.cryptoki/share/native/libj2pkcs11/pkcs11t.h:112]: These NSS specific defines is better off to be placed in a separate header file for ease of maintenance. In the past when PKCS#11 version is updated, we just drop the latest header files in and replace existing ones. Why not adding them to  to j2secmod.h or pkcs11wrapper.h?

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/Secmod.java [hunk 3]
```diff
@@ -392,20 +398,21 @@ public Module getModule(ModuleType type) {
...
+                    type = ModuleType.KEYSTORE;
+                } else if (slotId == FIPS_SLOT_ID) {
                     type = ModuleType.FIPS;
-                    if (slot != 0) {
-                        throw new RuntimeException
-                            ("Slot index should be 0 for FIPS slot");
-                    }
+                } else {
+                    throw new RuntimeException("Unexpected slot ID in the" +
+                            " NSS Internal Module");
```
Comment (valeriepeng) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/Secmod.java:415]: Add the slot ID to the exception message?

File: src/jdk.crypto.cryptoki/share/native/libj2pkcs11/j2secmod.h [hunk 4]
```diff
@@ -51,29 +51,137 @@ void *findFunction(JNIEnv *env, jlong jHandle, const char *functionName);
...
+typedef enum {
+    PK11_DIS_NONE = 0,
+    PK11_DIS_USER_SELECTED = 1,
+    PK11_DIS_COULD_NOT_INIT_TOKEN = 2,
+    PK11_DIS_TOKEN_VERIFY_FAILED = 3,
+    PK11_DIS_TOKEN_NOT_PRESENT = 4
+} PK11DisableReasons;
+
+/* represent a pkcs#11 slot reference counted. */
+struct PK11SlotInfoStr {
```
Comment (valeriepeng) [inline src/jdk.crypto.cryptoki/share/native/libj2pkcs11/j2secmod.h:78]: nit: add which nss header this is from.

File: src/jdk.crypto.cryptoki/share/native/libj2pkcs11/j2secmod.h [hunk 5]
```diff
@@ -51,29 +51,137 @@ void *findFunction(JNIEnv *env, jlong jHandle, const char *functionName);
...
+    void *nssToken;
+    /* the tokeninfo struct */
+    CK_TOKEN_INFO tokenInfo;
+    /* fast mechanism lookup */
+    char mechanismBits[256];
+    CK_PROFILE_ID *profileList;
+    int profileCount;
+};
+
 struct SECMODModuleStr {
```
Comment (valeriepeng) [inline src/jdk.crypto.cryptoki/share/native/libj2pkcs11/j2secmod.h:166]: Same nit: add which nss header this is from.

**Review by valeriepeng — COMMENTED** [review]
Here are some comments. Rest looks fine. Thanks. Valerie

File: src/jdk.crypto.cryptoki/share/native/libj2pkcs11/pkcs11t.h (hunk 1 above)
Comment (martinuy) [reply src/jdk.crypto.cryptoki/share/native/libj2pkcs11/pkcs11t.h:105]: oh, missed it. Thanks

File: src/jdk.crypto.cryptoki/share/native/libj2pkcs11/pkcs11t.h (hunk 2 above)
Comment (martinuy) [reply src/jdk.crypto.cryptoki/share/native/libj2pkcs11/pkcs11t.h:112]: Sure, moved to j2secmod.h

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/Secmod.java [hunk 6]
```diff
@@ -73,6 +73,12 @@
 
     private static final String TRUST_LIB_NAME = "nssckbi";
 
+    private final static int NETSCAPE_SLOT_ID = 0x1;
+
+    private final static int PRIVATE_KEY_SLOT_ID = 0x2;
+
+    private final static int FIPS_SLOT_ID = 0x3;
+
```
[the remark this answers is not in the document]
Comment (martinuy) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/Secmod.java:81]: Good, will add references to both

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/Secmod.java (hunk 3 above)
Comment (martinuy) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/Secmod.java:415]: Good idea

File: src/jdk.crypto.cryptoki/share/native/libj2pkcs11/j2secmod.h (hunk 4 above)
Comment (martinuy) [reply src/jdk.crypto.cryptoki/share/native/libj2pkcs11/j2secmod.h:78]: Good

File: src/jdk.crypto.cryptoki/share/native/libj2pkcs11/j2secmod.h (hunk 5 above)
Comment (martinuy) [reply src/jdk.crypto.cryptoki/share/native/libj2pkcs11/j2secmod.h:166]: Good. Done for every structure there

**Comment by martinuy** [comment]
@valeriepeng can you please take one more look at the changes?

No regressions found in jdk/sun/security/pkcs11 tests category.

**Review by valeriepeng — APPROVED** [review]
The updates look good. Thanks! Valerie

**Pull request closed without merging**